reciprocalspaceship/io/crystfel.py

```diff
@@ -0,0 +1,190 @@
+            elif is_cstar(line):
+                cstar = np.array(list(map(
+                    float,
+                    line.split()[2:5]))) / 10.  # crystfel's notation uses nm-1
+
+                # since it's the last line needed to construct Ewald offset,
+                # we'll pre-compute the matrices here
+                A = np.array([astar, bstar, cstar]).T
+                lambda_inv = 1 / eV2Angstrom(photon_energy)
+                s0 = np.array([0, 0, lambda_inv]).T
+
+            elif is_crystal:
+                # example line:
+                #    h    k    l          I   sigma(I)       peak background  fs/px  ss/px panel
+                #  -63   41    9     -41.31      57.45     195.00     170.86  731.0 1350.4 p0
+                crystal_peak_number += 1
+                h, k, l, I, sigmaI, _, _, _, _, _ = [i for i in line.split()]
+                h, k, l = map(float, [h, k, l])
+
+                # calculate ewald offset and s1
+                hkl = np.array([h, k, l])
+                # print(A)
+                s1 = A @ hkl + s0
+                s1x, s1y, s1z = s1
+                ewald_offset = np.linalg.norm(s1) - lambda_inv
+
+                record = {
+                    "H": h,
+                    "K": k,
+                    "L": l,
+                    "I": float(I),
+                    "sigmaI": float(sigmaI),
+                    "BATCH": crystal_idx,
+                    's1x': s1x,
+                    's1y': s1y,
+                    's1z': s1z,
+                    'ewald_offset': ewald_offset
+                }
+                if current_event is not None:
+                    name = (current_filename, current_event,
+                            current_serial_number, crystal_idx, crystal_peak_number)
+                else:
+                    name = (current_filename, current_serial_number, crystal_idx,
+                            crystal_peak_number)
+                answ_crystals[name] = record
+
+            # start analyzing where we are now
+            if corrupted_chunk:
+                if "Begin chunk" not in line:
+                    continue
+                else:
+                    is_crystal, is_chunk = False, False
+                    corrupted_chunk = False
+                    continue
+
+            if contains_filename(line):
+                current_filename = line.split()[-1]
+            elif contains_event(line):
+                current_event = line.split()[-1][2:]
+            elif contains_serial_number(line):
+                current_serial_number = line.split()[-1]
+
+            elif starts_chunk_peaks(line):
+                is_chunk = True
+                continue
+
+            elif starts_crystal_peaks(line):
+                crystal_idx += 1
+                is_crystal = True
+                continue
+
+    return answ_crystals
+
+
+def read_crystfel(streamfile, logfile=None) -> DataSet:
```

Comment:
Please remove `logfile=None` from the call signature. It may be useful to add optional `spacegroup` and `cell`, but we can discuss elsewhere

Reply:
Removed `logfile` and moved conversation about `spacegroup` and `cell` below